src/test/kotlin/logic/services/AuthenticationRepoImplTest.kt

```diff
@@ -0,0 +1,191 @@
+
+    private lateinit var hashingService: HashingService
+    private lateinit var authRepo: AuthenticationRepository
+    private lateinit var users: MutableList<User>
+
+    @BeforeTest
+    fun setup() {
+        hashingService = mockk()
+        users = mutableListOf()
+        authRepo = AuthenticationRepoImpl(hashingService, users)
+    }
+
+    @Test
+    fun `login should return user when credentials are valid`(){
+        val username = "abdo"
+        val rawPassword = "pass123"
+        val hashedPassword = "hashed-pass"
+        val storedUser = Mate(id= UUID.randomUUID(),username, hashedPassword)
+        users.add(storedUser)
+
+        val inputUser = Mate(id= UUID.randomUUID(),username, rawPassword)
+        every { hashingService.verifyData(rawPassword, hashedPassword) } returns true
+
+        val result = authRepo.login(inputUser)
+
+        assertEquals(storedUser, result)
+    }
+
+    @Test
+    fun `login should throw UserNotFoundException when user does not exist`() {
+        val username = "unknown"
+        val user = Mate(id= UUID.randomUUID(),username, "somepass")
+
+        val exception = assertFailsWith<UserNotFoundException> {
+            authRepo.login(user)
+        }
+
+        assertEquals("User 'unknown' not found", exception.message)
+    }
+
+    @Test
+    fun `login should throw InvalidPasswordException when password is incorrect`() {
+        val username = "ahmed"
+        val correctHashed = "hashed-pass"
+        val wrongInput = "wrong-pass"
+        val storedUser = Mate(id= UUID.randomUUID(),username, correctHashed)
+        users.add(storedUser)
+
+        val inputUser = Mate(id= UUID.randomUUID(),username, wrongInput)
+        every { hashingService.verifyData(wrongInput, correctHashed) } returns false
+
+        assertFailsWith<InvalidPasswordException> {
+            authRepo.login(inputUser)
+        }
+    }
+
+    @Test
+    fun `createMate should fail if user already exists`() {
+        val admin = Admin(id= UUID.randomUUID(),"admin1", "hashed123")
+        val existingUsername = "mate1"
+        val newMate= Mate(id= UUID.randomUUID(),existingUsername, "pass")
+        users.add(newMate)
+
+        assertFailsWith<UserAlreadyExistException> {
+            authRepo.createMate(admin, existingUsername, "newPass")
+        }
+    }
+
+    @Test
+    fun `createMate should pass when admin is the creator`() {
+        val admin = Admin(id= UUID.randomUUID(),"admin1", "hashed123")
+        val username = "newMate"
+        val rawPassword = "password"
+        val expectedHash = "hashedPassword"
+
+        every { hashingService.hashData(rawPassword) } returns expectedHash
+
+        val mate = authRepo.createMate(admin, username, rawPassword)
+
+        assertEquals(username, mate.username)
+        assertEquals(expectedHash, mate.password)
+        verify(exactly = 1) { hashingService.hashData(rawPassword) }
+
+        assertTrue(users.contains(mate))
```

Comment:
This can be split into two test cases, one for the created mate credentials, and another for asserting that the mate is created.